db/db.go

```diff
@@ -0,0 +1,260 @@
+func (m *mongo) Connect() error {
+	session, err := mgo.Dial(m.url)
+	if err != nil {
+		return ErrConnect
+	}
+
+	m.session = session
+	return nil
+}
+
+// Disconnect must be called when finished with db connector.
+func (m *mongo) Disconnect() {
+	m.session.Close()
+}
+
+// CreateEntities inserts many Entity into DB.
+func (m *mongo) CreateEntities(entities []Entity) []string {
+	s := m.session.Copy()
+	defer s.Close()
+
+	c := m.session.DB(m.database).C(m.collection)
+
+	var objectIDs []string
+
+	for _, e := range entities {
+		objectID := bson.NewObjectId().String()
+		objectIDs = append(objectIDs, objectID)
+		// Note: this modifies entity in memory
+		e["_id"] = objectID
+
+		c.Insert(e)
+	}
+
+	return objectIDs
+}
+
+// Query mongoDB instance with a Query object, which is tranlsated to bson.M
+// object for mgo driver to understand. Returns a list of Entity objects. If
+// nothing is found, returns empty list.
+func (m *mongo) ReadEntities(q query.Query) ([]Entity, error) {
+	mgoQuery := translateQuery(q)
+
+	s := m.session.Copy()
+	defer s.Close()
+
+	var entities []Entity
+	c := s.DB(m.database).C(m.collection)
+	err := c.Find(mgoQuery).All(&entities)
+	if err != nil {
+		return nil, ErrNotFound
+	}
+
+	return entities, nil
+}
+
+// UpdateEntities does a db.collection.findAndModify() operation
+// (https://docs.mongodb.com/manual/reference/method/db.collection.findAndModify/)
+// on all docs returned from query. Returns diff of entities (old values for
+// fields changed, in addition to including "_id" field/value) or an error, if
+// one occured.
+func (m *mongo) UpdateEntities(q query.Query, u Entity) ([]Entity, error) {
+	mgoQuery := translateQuery(q)
+
+	s := m.session.Copy()
+	defer s.Close()
+
+	c := s.DB(m.database).C(m.collection)
+
+	// List of IDs for docs to update
+	ids, err := idsForQuery(c, mgoQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	// Change to make
+	change := mgo.Change{Update: bson.M{"$set": u}}
+
+	// Query for each document and apply update
+	var diffs []Entity
+	for _, id := range ids {
+		var diff Entity
+		_, err := c.Find(bson.M{"_id": id}).Select(selectMap(u)).Apply(change, &diff)
+		if err != nil {
+			return nil, ErrUpdateEntity
+		}
+
+		diffs = append(diffs, diff)
+	}
+
+	return diffs, nil
+}
+
+// Deletes many Entity objects from DB. Returns old entities that were removed
+// or an error, if one occured.
+func (m *mongo) DeleteEntities(q query.Query) ([]Entity, error) {
+	mgoQuery := translateQuery(q)
+
+	s := m.session.Copy()
+	defer s.Close()
+
+	c := s.DB(m.database).C(m.collection)
+
+	// List of IDs for docs to update
+	ids, err := idsForQuery(c, mgoQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	// Change to make
+	change := mgo.Change{Remove: true}
+
+	// Query for each document and delete it
+	var results []Entity
+	for _, id := range ids {
+		var result Entity
+		_, err := c.Find(bson.M{"_id": id}).Apply(change, &result)
```

Comment:
Or, `var deletedEntity Entity` might preclude the need for a code comment because then the code is more intention-revealing: I can put 2 and 2 together and see that `deletedEntity` is most likely the deleted entity which gets returned later (whereas `return result` could mean result of the operation).